### DISCUSSION TITLE
ZSA Voyager (or how to deal with one less thumb?)

### DISCUSSION THREAD

**Discussion by TaiSHiNet** [manna-harbour/miryoku#241] [Q&A] [answered]
Hi all,

Just saw the announcement for the ZSA Voyager and, while it's beautiful, it lacks the extra thumb on each side.
What would be a "correct" (using the term very loosely here) adaptation from Miryoku?
I guess H/D could (or V/K) be used for mods depending on how well the thumbs reach them

Looking forward to hearing your thoughts!

**Comment by manna-harbour** [comment]
- https://github.com/manna-harbour/miryoku/issues/56 
- https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#thumb-combos

I'll add support for Voyager to Miryoku QMK soon.

**Reply by TaiSHiNet** [reply]
That's amazing, thanks! Also thank you for your hard work

**Comment by manna-harbour** [comment, marked as answer]
Voyager will be added to [Miryoku Oryx](https://github.com/manna-harbour/miryoku/tree/master/data/oryx) when time permits, using the native hand position and [thumb combos](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#thumb-combos).

Voyager is not yet in QMK master, so Voyager won't yet be added to Miryoku QMK in the `miryoku` branch, but has been added in a temporary branch. Voyager for QMK is at https://github.com/zsa/qmk_firmware/tree/firmware22/keyboards/voyager and Voyager for Miryoku QMK is at https://github.com/manna-harbour/miryoku_qmk/tree/voyager-test/users/manna-harbour_miryoku#voyager. The ZSA fork of QMK is too different to merge, so it can't be built using the [Miryoku QMK workflow merge branch feature](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#build-inputs). You'll have to merge manually or copy files until ZSA make a PR to QMK.

The more comfortable alternative layouts included for [Ergodox in Miryoku QMK](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#ergodox) won't be possible due to lack of keys. However `MIRYOKU_MAPPING=SHIFTED_ROWS` is possible and has been included, and will probably be more comfortable than the native hand position and thumb combos, included as the default.

- [x] Create test branch for use with ZSA fork
- [x] Update test branch when PR is available: https://github.com/manna-harbour/miryoku/discussions/241#discussioncomment-7641111
- [ ] Add to `miryoku` branch when PR is merged
- [ ] Add Voyager to Miryoku Oryx

**Reply by manna-harbour** [reply]
@drashna has now opened a PR for Voyager in QMK: https://github.com/qmk/qmk_firmware/pull/22181. The PR branch is https://github.com/drashna/qmk_firmware/tree/keyboard/zsa_voyager. The keyboard directory has changed from `keyboards/voyager` to `keyboards/zsa/voyager`. The `voyager-test` branch above has been adjusted to match.

To build, merge the `voyager-test` branch and the PR branch. If using the Miryoku QMK build workflows, select the `voyager-test` branch in the workflow branch selector, use `zsa/voyager` with the keyboard field, and use `drashna/qmk_firmware/keyboard/zsa_voyager` with the merge field. Here's a sample build run: https://github.com/manna-harbour/miryoku_qmk/actions/runs/6956266218.

**Reply by davidsteinberger** [reply]
This is working great. The only issue that I have, is, that caps lock does not work for me. Caps word works. @manna-harbour any idea what could be causing this?

**Reply by elisiano** [reply]
@manna-harbour in case you'd like a starting point for the Miryoku Oryx, I've just finished my first Voyager implementation, you can find it [here](https://configure.zsa.io/voyager/layouts/WGGgl/LX7ey/0).

It's not super complete compared to the qmk implementation, but I used your moonlander layout as reference (also, I implemented as shifted rows because I needed space for the thumbs).

**Comment by alexwaibel** [comment]
Took a stab at this by shifting the keys up one and using one of the bottom row keys as the missing thumb key.
https://configure.zsa.io/voyager/layouts/g9gZ4/latest/0

**Reply by manna-harbour** [reply]
Thanks for sharing! That might be the best option, depending on the thumb keys position. Let us know after the board arrives.

If that makes the 2 thumbs keys too far away, try a [shifted rows mapping](https://raw.githubusercontent.com/manna-harbour/miryoku/master/data/mapping/miryoku-kle-mapping-ergodox-shifted_rows.png), or keep the native hand position and add [thumb combos](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#thumb-combos).

**Reply by alexwaibel** [reply]
I actually have the board already! With the layout I posted, I can comfortably reach all the layer buttons but some users may find the lowest thumb buttons difficult to reach and could maybe substitute v and k as thumb buttons or use shifted rows. Docs on thumb combos are a little sparse, is the idea to hold both thumb buttons on one hand or one from each hand? I played around with combos in oryx a bit but wasn't sure where a good spot to put esc and delete would be. Maybe next to home row pinky?

**Reply by alexwaibel** [reply]
Here's a version with thumb combos as I understand them. Funny enough, it crashes the zsa keymapp software when previewing the layout and pressing the combo. Working well so far but definitely open to suggestions.
https://configure.zsa.io/voyager/layouts/xLzJo/latest/3

**Reply by armax00** [reply]
I am doing the same, and also got the device. This is my layout variation: https://configure.zsa.io/voyager/layouts/0pP9N/latest/0 and it feels quite comfortable. I have the same setup for the moonlander, so to not have the feeling of sitting on two too different keyboards: https://configure.zsa.io/moonlander/layouts/G573V/latest/0

**Reply by alexwaibel** [reply]
@armax00 your layouts seem to be private btw. To anyone else using thumb combos, where did you put esc and delete? I find they're usable on the outer pinkie column but not quite as comfortable as having them on a thumb button. Makes switching back and forth with the moonlander layout a bit confusing as well.

**Reply by manna-harbour** [reply]
> is the idea to hold both thumb buttons on one hand or one from each hand?

Combo the primary and secondary thumb keys to emulate the tertiary thumb key, all on the same hand. The combo can be tapped or held, and has the same function as tapping or holding the tertiary thumb key, on all layers.

> To anyone else using thumb combos, where did you put esc and delete? I find they're usable on the outer pinkie column but not quite as comfortable as having them on a thumb button. 

Thumb combos cover taps and holds on all layers, so include Escape and Delete.

> Makes switching back and forth with the moonlander layout a bit confusing as well.

You can also add thumb combos to other keyboards that don't require it, for consistency.

**Reply by alexwaibel** [reply]
Thanks for the detailed follow up. I've just been using oryx and their combo support seems to be more limited so I don't think I can differentiate combo taps vs holds. Guess it's time to dive into your QMK builds.

**Reply by armax00** [reply]
@alexwaibel I believe I simply copied the wrong URLs. I updated them, can you see the layouts now?

**Reply by manna-harbour** [reply]
> I've just been using oryx and their combo support seems to be more limited so I don't think I can differentiate combo taps vs holds. 

The trick in general is to just have the combos map to the same keycodes as the tertiary thumb. On Base that's a tap-hold, so that's where the distinction is made, not by the combo itself. See the QMK code from here for reference: https://github.com/manna-harbour/miryoku_qmk/blob/e1307cc3a71e8f23c6f6c337764d6c854da009a6/users/manna-harbour_miryoku/manna-harbour_miryoku.c#L62

**Reply by davidsteinberger** [reply]
@alexwaibel @armax00 I've just ordered the voyager too. While waiting I wanted to dive into layout/firmware details.
What did u settle on? Any recommendations?

**Reply by armax00** [reply]
@davidsteinberger I am keeping on with my variation of layout with the keys shifted up (which is quite ok). I have also removed some keys with the ZipKit I have. Note also that since I use VIM I have adjusted the navigation keys to where they are in VIM in Colemak-DH. It is a transition but I am getting there :)

**Reply by davidsteinberger** [reply]
Thanks! I've just received mine and flashed miryoku qmk onto it. Works very nice, but does anybody know how I can get the top row to work (currently it's dead)?

**Reply by davidsteinberger** [reply]
Answering my last question: this was actually rather easy. I just defined the _MIRYOKU_MAPPING_ in my _custom_config.h_. Not sure what I will eventually use this row for, but for now I'm super happy with how things are. If somebody is interested in it or needs help compiling qmk let me know.

**Reply by martinezj** [reply]
@davidsteinberger I'm receiving my voyager next week and I've tried several times to compile the QMK file and I can't get it to work. If you don't mind, what input string did you use to make the build. That might help me figure out what I'm doing wrong. Thx

**Reply by mitcdh** [reply]
@manna-harbour Unless I'm mistaken, unfortunately, Oryx doesn't support tap/hold distinction for combos, so the following wouldn't work:

````
COMBO(thumbcombos_base_right, LT(U_FUN, KC_DEL)),
COMBO(thumbcombos_base_left, LT(U_MEDIA, KC_ESC)),
````

You can have it map to U_FUN or KC_DEL but not both; it also appears only to support tap rather than hold operations, so OSL/TO rather than MO. Unless there's a creative way of mapping to another key that I'm missing? It appears all kludge combos work other than these and the middle mouse (where Oryx doesn't let you select KC_BTN2 and KC_BTN1 as part of a combo).

Here's a test layout which will do OSL+(ESC|Del) on either of the thumb keys: https://configure.zsa.io/voyager/layouts/Jz03z/latest/0

**Reply by manna-harbour** [reply]
@davidsteinberger
> Thanks! I've just received mine and flashed miryoku qmk onto it. Works very nice, but does anybody know how I can get the top row to work (currently it's dead)?

You've found the answer, but for others, see https://github.com/manna-harbour/miryoku/discussions/85.

**Reply by manna-harbour** [reply]
@mitcdh
> Oryx doesn't support tap/hold distinction for combos

Thanks for the info! In that case the Miryoku Oryx version will have to use shifted rows instead, like the [Miryoku QMK option](https://github.com/manna-harbour/miryoku_qmk/tree/voyager-test/users/manna-harbour_miryoku#voyager).

**Reply by mitcdh** [reply]
This is the source that Oryx generates:

```
const uint16_t PROGMEM combo0[] = { KC_ENTER, KC_BSPACE, COMBO_END};
const uint16_t PROGMEM combo1[] = { KC_MEDIA_PLAY_PAUSE, KC_MEDIA_STOP, COMBO_END};
const uint16_t PROGMEM combo2[] = { KC_KP_MINUS, KC_0, COMBO_END};
const uint16_t PROGMEM combo3[] = { KC_RPRN, KC_UNDS, COMBO_END};
const uint16_t PROGMEM combo4[] = { KC_TAB, KC_SPACE, COMBO_END};
const uint16_t PROGMEM combo5[] = { LT(5,KC_ENTER), LT(4,KC_BSPACE), COMBO_END};
const uint16_t PROGMEM combo6[] = { LT(1,KC_SPACE), LT(2,KC_TAB), COMBO_END};

combo_t key_combos[COMBO_COUNT] = {
    COMBO(combo0, KC_DELETE),
    COMBO(combo1, KC_AUDIO_MUTE),
    COMBO(combo2, KC_KP_DOT),
    COMBO(combo3, KC_LPRN),
    COMBO(combo4, KC_APPLICATION),
    COMBO(combo5, OSL(6)),
    COMBO(combo6, OSL(3)),
};
```

So I imagine it's still possible to do without shifted rows, maybe just not through an Oryx configurator-compatible web distribution.

**Reply by manna-harbour** [reply]
@mitcdh 
> I imagine it's still possible to do without shifted rows, maybe just not through an Oryx configurator-compatible web distribution

Shifted rows is this:

https://github.com/manna-harbour/miryoku_qmk/tree/voyager-test/users/manna-harbour_miryoku#voyager
> An alternative subset mapping is provided with all keys shifted up one row creating thumb keys in the original alpha area. To select this mapping, append MIRYOKU_MAPPING=SHIFTED_ROWS to the make command line when building.

That's an option in Miryoku QMK for voyager (and [ergodox](https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#ergodox), and will be added to moonlander by request). That can also be done in Oryx, and will be used in Miryoku Oryx for voyager if Oryx lacks full combo support.

**Reply by manna-harbour** [reply]
@martinezj
> I've tried several times to compile the QMK file and I can't get it to work.

Due to https://github.com/manna-harbour/miryoku/discussions/287, automatic merging in the workflow as described at https://github.com/manna-harbour/miryoku/discussions/241#discussioncomment-7641111 no longer works. You'll have to manually copy https://github.com/drashna/qmk_firmware/tree/keyboard/zsa_voyager/keyboards/zsa/voyager into https://github.com/manna-harbour/miryoku_qmk/tree/voyager-test/keyboards/zsa/voyager.

**Reply by mitcdh** [reply]
> @mitcdh
> 
> > Oryx doesn't support tap/hold distinction for combos
> 
> Thanks for the info! In that case the Miryoku Oryx version will have to use shifted rows instead, like the [Miryoku QMK option](https://github.com/manna-harbour/miryoku_qmk/tree/voyager-test/users/manna-harbour_miryoku#voyager).

I have the kludge combos working with this: https://configure.zsa.io/voyager/layouts/pzGDD/rL7Pb/0

**Comment by davidsteinberger** [comment]
ZSA's fork is quite a bit behind, so I went the copy/paste route. Here's what I did.

```
git clone https://github.com/zsa/qmk_firmware.git
git clone https://github.com/manna-harbour/miryoku_qmk.git
cd ./miryoku_qmk
git checkout voyager-test
cd ../qmk_firmware
git checkout firmware23
mkdir -p ./users/manna-harbour_miryoku
mkdir -p ./keyboards/voyager/keymaps/manna-harbour_miryoku
cp -r ../miryoku_qmk/users/manna-harbour_miryoku ./users
cp -r ../miryoku_qmk/keyboards/voyager/keymaps/manna-harbour_miryoku ./keyboards/voyager/keymaps
make git-submodule
```
manually delete line 15 in `./users/manna-harbour_miryoku/config.h` (`#define IGNORE_MOD_TAP_INTERRUPT`)
@manna-harbour is there any other way to do it?
```
make clean voyager:manna-harbour_miryoku
```

**Reply by martinezj** [reply]
@davidsteinberger I can't thank you enough for the step by step, this is exactly what I needed.  I went through the steps and it worked perfectly, I just loaded the firmware on the Voyager and it's working great so far. Thanks again for your help with this. @manna-harbour if you have any additional advise it would also be greatly appreciated.

**Reply by martinezj** [reply]
@davidsteinberger if it's not too much to ask could I see your custom_config.h. file to see an example of how you added a mapping for the top row. I'm so happy to have the board working after the help you gave me.

**Reply by davidsteinberger** [reply]
Here is a simplified version of mine. I removed some changes for combos that I have as well as they also require changes to `manna-harbour_miryoku.c`.

[custom_config.h.txt](https://github.com/manna-harbour/miryoku/files/13446802/custom_config.h.txt)

**Reply by martinezj** [reply]
Perfect! Thanks again.

**Reply by manna-harbour** [reply]
Also see https://github.com/manna-harbour/miryoku/discussions/241#discussioncomment-7641111.

**Comment by BugRoger** [comment]
Here's a working Miryoku build based on the new qmk_userspace. It builds a default Miryoku as well as my customized keymap with Achordion and shifted rows. Super easy to build with the Github actions. 

https://github.com/BugRoger/qmk_userspace/
